gordon_janitor_gcp/reconciler.py

```diff
@@ -0,0 +1,243 @@
+            action (str): (optional) action for these corrective
+                messages. Defaults to ``'additions'``.
+        """
+        for rrset in desired_rrsets:
+            msg = {
+                'resourceRecords': attr.asdict(rrset),
+                'action': action
+            }
+            # TODO (lynn): add metrics.incr call here once aioshumway is
+            #              released
+            logging.debug('Creating the following change message: {msg}')
+            await self.changes_channel.put(msg)
+
+        logging.info(f'Created {len(desired_rrsets)} change messages.')
+
+    def _parse_rrset_message(self, message):
+        # assert that keys 'zone' and 'rrsets' are present, and return
+        # values for each
+        try:
+            zone = message['zone']
+        except KeyError:
+            msg = (f'No zone was defined in the given message: {message}.')
+            logging.error(msg)
+            raise exceptions.GCPGordonJanitorError(msg)
+        try:
+            rrsets = message['rrsets']
+        except KeyError:
+            msg = (f'No resource record sets were defined in given message: '
+                   f'{message}.')
+            logging.error(msg)
+            raise exceptions.GCPGordonJanitorError(msg)
+
+        return zone, rrsets
+
+    async def validate_rrsets_by_zone(self, zone, rrsets):
+        """Given a zone, validate current versus desired rrsets.
+
+        If there are any missing records, a corrective message for each
+        record will be published.
+
+        Args:
+            zone (str): zone to query Google Cloud DNS API.
+            rrsets (list): desired record sets to which to compare the
+                Cloud DNS API's response.
+        """
+        desired_rrsets = [
+            cloud_dns.GCPResourceRecordSet(**record) for record in rrsets
+        ]
+
+        actual_rrsets = await self.dns_client.get_records_for_zone(zone)
+
+        # NOTE: only working on "additions" (which includes changes to
+        #       current records) right now.
+        # TODO: (FEATURE) add support for cleaning up records that
+        #       should have been deleted.
+        missing_rrsets = [
+            rs for rs in desired_rrsets if rs not in actual_rrsets
+        ]
+        await asyncio.ensure_future(
+            self.publish_change_messages(missing_rrsets, action='additions'))
+
+    async def start(self):
+        """Start consuming from :py:obj:`self.rrset_channel`.
+
+        Once ``None`` is received from the channel, finish processing
+        records and emit a ``None`` message to the
+        :py:obj:`self.changes_channel`.
+        """
+        while True:
+            desired_rrset = await self.rrset_channel.get()
+            if desired_rrset is None:
+                break
+            try:
+                zone, raw_rrsets = self._parse_rrset_message(desired_rrset)
+                await asyncio.ensure_future(
```

Comment:
What's the intent behind using `await asyncio.ensure_future`?

If I understand the flow correctly, it will process each incoming message sequentially and then it will wait up to 60 seconds (default timeout) for the last incoming message to be processed before pushing `None` to the `changes_channel`. Is this about right?